test/logger-errors.js

```diff
@@ -0,0 +1,385 @@
+      return true
+    })
+    .query(() => { return true })
+    .reply(200, 'Ingester response')
+    .persist()
+
+  logger.on('send', (obj) => {
+    t.deepEqual(obj, {
+      httpStatus: 200
+    , firstLine: 'my log line'
+    , lastLine: 'my log line'
+    , totalLinesSent: 2
+    , totalLinesReady: 0
+    , bufferCount: 0
+    }, 'Got send event')
+  })
+
+  logger.info('my log line', {
+    timestamp: {}
+  })
+  logger.info('my log line', {
+    timestamp: startTime + constants.MS_IN_A_DAY
+  })
+})
+
+test('.log() throws if options is a string and not a valid log entry', (t) => {
+  t.plan(2)
+  const logger = new Logger(apiKey, createOptions())
+
+  t.on('end', async () => {
+    nock.cleanAll()
+  })
+
+  nock(logger.url)
+    .post('/', () => { return true })
+    .query(() => { return true })
+    .reply(200, 'Ingester response')
+
+  logger.on('error', (err) => {
+    t.type(err, TypeError, 'Expected to be a TypeError')
+    t.match(err, {
+      name: 'TypeError'
+    , message: 'If \'options\' is a string, then it must be a valid log level'
+    , meta: {
+        got: 'NOPE'
+      , expected: constants.LOG_LEVELS
+      }
+    }, 'Expected Error is correct')
+  })
+  logger.log('log line', 'NOPE')
+})
+
+test('.log() rejects invalid `opts` data type', (t) => {
+  t.plan(2)
+  const logger = new Logger(apiKey, createOptions())
+
+  t.on('end', async () => {
+    nock.cleanAll()
+  })
+
+  nock(logger.url)
+    .post('/', () => { return true })
+    .query(() => { return true })
+    .reply(200, 'Ingester response')
+
+  logger.on('error', (err) => {
+    t.type(err, TypeError, 'Expected to be a TypeError')
+    t.match(err, {
+      name: 'TypeError'
+    , message: 'options parameter must be a level (string), or object'
+    , meta: {
+        got: 'number'
+      }
+    }, 'Expected Error is correct')
+  })
+  logger.log('log line', 12345)
+})
+
+test('.log() passed an invalid log level uses the default instead', (t) => {
+  t.plan(3)
+  const logger = new Logger(apiKey, createOptions({
+    level: 'info'
+  }))
+
+  t.on('end', async () => {
+    nock.cleanAll()
+  })
+
+  nock(logger.url)
+    .post('/', (body) => {
+      const obj = body.ls[0]
+      t.equal(obj.level, 'INFO', 'Default level was used instead')
+      return true
+    })
+    .query(() => { return true })
+    .reply(200, 'Ingester response')
+
+  logger.on('error', (err) => {
+    t.type(err, Error, 'Expected to be an Error')
+    t.match(err, {
+      name: 'Error'
+    , message: 'Invalid log level.  Using the default instead.'
+    , meta: {
+        got: 'NEIGH'
+      , expected: constants.LOG_LEVELS
+      , used: logger.level
+      }
+    }, 'Expected Error is correct')
+  })
+  logger.log('log line', {level: 'NEIGH'})
+})
+
+test('.log() ignores empty or null messages', (t) => {
+  t.plan(6)
+
+  const logger = new Logger(apiKey, createOptions())
+  const expected = [null, undefined, '']
+
+  logger.on('warn', (obj) => {
+    t.match(obj, {
```

Comment:
deepEqual

Reply:
`deepEqual` is good here, too.  It was missing `statement: ''`, which should definitely be asserted here.

Reply:
ok, so this one is tricky.  The same assertion is used for all values on line 166, so the value differs.  I agree that this should test `statement`, so this will have to be broken up into subtests since that's simpler than trying to rely on asynchronous events being ordered (and the complexity for changing the value).

Reply:
Broken out into subtests to be explicit